In [1]:
import pandas as pd
from collections import Counter
import datetime
import numpy as np
import warnings
import time
%matplotlib inline

warnings.filterwarnings('ignore')

In [2]:
users = pd.read_csv('../Data/raw_data_b/user_register_log.txt', header=None,names=['user_id','register_day','register_type','device_type'],sep='\t') # 注册用户数据加载
user_login = pd.read_csv('../Data/raw_data_b/app_launch_log.txt', header=None,names=['user_id','day'],sep='\t') # app登录日志数据加载
user_act = pd.read_csv('../Data/raw_data_b/user_activity_log.txt', header=None,names=['user_id','day','page','video_id','author_id','action_type'],sep='\t') # 用户行为日志数据加载
user_video = pd.read_csv('../Data/raw_data_b/video_create_log.txt', header=None,names=['user_id','day'],sep='\t') # 用户拍摄视频日志数据加载

所有数据都是1-30天内的数据

且无缺失值

# 对数据进行滑窗法划分：

data: 

    +-data1
        --feature:1-15;label:16-22
    +-data2
        --feature:9-23;label:24-30
    +-data3
        --feature:1-30;label: 31-37

In [3]:
def cutDataFunc(data, cut_col ,start_day, end_day):
    return data[(data[cut_col]<=end_day)&(data[cut_col]>=start_day)]

def cutDataByTime(data_url, start_day, end_day):
    temp_users = cutDataFunc(users, 'register_day', start_day, end_day)
    temp_login = cutDataFunc(user_login, 'day', start_day, end_day)
    temp_act = cutDataFunc(user_act, 'day', start_day, end_day)
    temp_video = cutDataFunc(user_video, 'day', start_day, end_day)
    
    temp_users.to_csv(data_url+'users.csv',index=False)
    temp_login.to_csv(data_url+'login.csv',index=False)
    temp_act.to_csv(data_url+'act.csv',index=False)
    temp_video.to_csv(data_url+'video.csv',index=False)
 

In [4]:
def cutDataProgram():
    start = time.time()
    print ("---------START-----------")
    
    cutDataByTime('../Data/data1/train_', 1, 15)
    cutDataByTime('../Data/data1/test_', 16, 22)
    print ("-----第1数据集完成-------")
    
    cutDataByTime('../Data/data2/train_', 9, 23)
    cutDataByTime('../Data/data2/test_', 24, 30)
    print ("-----第2数据集完成-------")
    
    cutDataByTime('../Data/data3/train_', 1, 30)
    print ("-----第3数据集完成-------")
    
    print ("----------END------------")
    end = time.time()
    print (end - start)

In [5]:
cutDataProgram()

---------START-----------
-----第1数据集完成-------
-----第2数据集完成-------
-----第3数据集完成-------
----------END------------
238.051616191864


# train集构造和标签

In [2]:
# 获取有活跃行为的用户集，（从test集中获取）
def getActivityUsers(data_url):
    test_login = pd.read_csv(data_url+'test_login.csv')
    test_act = pd.read_csv(data_url+'test_act.csv')
    test_video = pd.read_csv(data_url+'test_video.csv')
    
    activity_user = np.unique(pd.concat([test_login['user_id'], test_act['user_id'], test_video['user_id']]))
    return activity_user

def get_diff_from_ls(x):
    x.sort()
    return list(np.diff(x))

# （video and lanuch）data create feature method
# 1、将day转换成距离时间窗口截点的距离
# 2、描述性day、cnt统计特征
# 3、连续1、 2、 3、 7天内的统计特征
def getCountFeature(data, name='login'):
    day_max = max(data['day'])
    data['day'] = day_max - data.day
    df = data.groupby(['user_id','day'],as_index=False).apply(lambda x:x.shape[0])
    df_temp = pd.DataFrame(df, columns=['cnt']).reset_index()
    
    df_temp_group = df_temp.groupby(['user_id'],as_index=False)
    res_df = df_temp_group.agg({'day':['max','min','std'],'cnt':['count','sum','max','var','mean']})
    res_df.columns = ['user_id', name+'_day_max',name+'_day_min',name+'_day_std',name+'_cnt',name+'_sum',name+'_max',name+'_var',name+'_mean']
    
    # 日期差分统计特征
#     df_temp_diff = df_temp[['user_id','day']].groupby('user_id').aggregate(lambda x: list(set(x)))
#     df_temp_diff.day = df_temp_diff.day.apply(lambda x:get_diff_from_ls(x))
#     res_df[name+'_day_diff_max'] = df_temp_diff.day.apply(lambda x: max(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_min'] = df_temp_diff.day.apply(lambda x: min(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_mean'] = df_temp_diff.day.apply(lambda x: np.mean(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_std'] = df_temp_diff.day.apply(lambda x: pd.Series(x).std() if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_skew'] = df_temp_diff.day.apply(lambda x: pd.Series(x).skew()).fillna(-1).values
#     res_df[name+'_day_diff_kurt'] = df_temp_diff.day.apply(lambda x: pd.Series(x).kurt()).fillna(-1).values
    
    
    for day_len in [1,3,7]:
        df_temp_day = df_temp[(df_temp.day>=0) & (df_temp.day<day_len)]
        df_temp_day_cnt = df_temp_day.groupby(['user_id']).apply(lambda x:sum(x['cnt'].values))
        add_df = pd.DataFrame(df_temp_day_cnt, columns=[name+'_'+str(day_len)+'_cnt']).reset_index()
        res_df = pd.merge(res_df, add_df, on='user_id', how='left').fillna(0)
        if day_len != 1:
            res_df[name+'_'+str(day_len)+'_cnt_arg'] = res_df[name+'_'+str(day_len)+'_cnt'] / day_len
            
    res_df['login_cnt_weight_sum'] = 0
    if name == 'login':
        for day_len in [1,3,7]:
            res_df['login_cnt_weight_sum'] += res_df[name+'_'+str(day_len)+'_cnt']*(8-day_len)
    return res_df.fillna(0)

# act: create feature method
# 1、将day转换成距离时间窗口截点的距离
# 2、描述性day、cnt统计特征
# 3、连续1、 2、 3、 7天内的统计特征
# 4、page和action_type不同类型的所有统计数以及比率特征
# 5、page和action_type不同类型在连续1、 3、 7天内的统计特征
# 6、user_id是否为author_id的成员
def getCountFeatureAboutAct(data, name='act'):
    day_max = max(data['day'])
    data['day'] = day_max - data.day
    authors = set(data['author_id'])
    
    df = data.groupby(['user_id','day'],as_index=False).apply(lambda x:x.shape[0])
    df_temp = pd.DataFrame(df, columns=['cnt']).reset_index()
    
    df_temp_group = df_temp.groupby(['user_id'],as_index=False)
    res_df = df_temp_group.agg({'day':['max','min','std'],'cnt':['count','sum','max','var','mean']})
    res_df.columns = ['user_id', name+'_day_max',name+'_day_min',name+'_day_std',name+'_cnt',name+'_sum',name+'_max',name+'_var',name+'_mean']
    
    # 日期差分统计特征
#     df_temp_diff = df_temp[['user_id','day']].groupby('user_id').aggregate(lambda x: list(set(x)))
#     df_temp_diff.day = df_temp_diff.day.apply(lambda x:get_diff_from_ls(x))
#     res_df[name+'_day_diff_max'] = df_temp_diff.day.apply(lambda x: max(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_min'] = df_temp_diff.day.apply(lambda x: min(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_mean'] = df_temp_diff.day.apply(lambda x: np.mean(x) if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_std'] = df_temp_diff.day.apply(lambda x: pd.Series(x).std() if len(x)!=0 else -1).fillna(-1).values
#     res_df[name+'_day_diff_skew'] = df_temp_diff.day.apply(lambda x: pd.Series(x).skew()).fillna(-1).values
#     res_df[name+'_day_diff_kurt'] = df_temp_diff.day.apply(lambda x: pd.Series(x).kurt()).fillna(-1).values
    
    for day_len in [1,3,7]:
        df_temp_day = df_temp[(df_temp.day>=0) & (df_temp.day<day_len)]
        df_temp_day_cnt = df_temp_day.groupby(['user_id']).apply(lambda x:sum(x['cnt'].values))
        add_df = pd.DataFrame(df_temp_day_cnt, columns=[name+'_'+str(day_len)+'_cnt']).reset_index()
        res_df = pd.merge(res_df, add_df, on='user_id', how='left').fillna(0)
        if day_len != 1:
            res_df[name+'_'+str(day_len)+'_cnt_arg'] = res_df[name+'_'+str(day_len)+'_cnt'] / day_len
    
    # page 处理
    temp = data.groupby(['user_id','day','page'],as_index=False).apply(lambda x:x.shape[0]).unstack().reset_index().fillna(0)
    for day_len in [1,3,7]:
        for col in [0,1,2,3,4]:
            temp_day = temp[(temp.day>=0) & (temp.day<day_len)]
            df_temp_day_cnt = temp_day.groupby(['user_id']).apply(lambda x:sum(x[col].values))
            add_df = pd.DataFrame(df_temp_day_cnt, columns=['page_'+str(col)+'_'+str(day_len)+'_cnt']).reset_index()
            res_df = pd.merge(res_df, add_df, on='user_id', how='left').fillna(0)
            if day_len != 1:
                res_df['page_'+str(col)+'_'+str(day_len)+'_cnt_arg'] = res_df['page_'+str(col)+'_'+str(day_len)+'_cnt'] / day_len
    
    page = data.groupby(['user_id','page'],as_index=False).apply(lambda x:x.shape[0]).unstack().reset_index().fillna(0)
    page_sum = page[0]+page[1]+page[2]+page[3]+page[4]
    res_df['page_0_sigle'] = page[0] / page_sum
    res_df['page_1_sigle'] = page[1] / page_sum
    res_df['page_2_sigle'] = page[2] / page_sum
    res_df['page_3_sigle'] = page[3] / page_sum
    res_df['page_4_sigle'] = page[4] / page_sum
#     res_df['page_sum'] = page_sum
    res_df[['page_0','page_1','page_2','page_3','page_4']] = page[[0,1,2,3,4]]
    
    # action_type处理
    temp = data.groupby(['user_id','day','action_type'],as_index=False).apply(lambda x:x.shape[0]).unstack().reset_index().fillna(0)
    temp_group_action = temp.groupby(['user_id'])
    for day_len in [1,3,7]:
        for col in [0,1,2,3,4,5]:
            temp_day = temp[(temp.day>=0) & (temp.day<day_len)]
            df_temp_day_cnt = temp_day.groupby(['user_id']).apply(lambda x:sum(x[col].values))
            add_df = pd.DataFrame(df_temp_day_cnt, columns=['action_type_'+str(col)+'_'+str(day_len)+'_cnt']).reset_index()
            res_df = pd.merge(res_df, add_df, on='user_id', how='left').fillna(0)
            if day_len != 1:
                res_df['action_type_'+str(col)+'_'+str(day_len)+'_cnt_arg'] = res_df['action_type_'+str(col)+'_'+str(day_len)+'_cnt'] / day_len
    
    action_type = data.groupby(['user_id','action_type'],as_index=False).apply(lambda x:x.shape[0]).unstack().reset_index().fillna(0)
    action_type_sum = action_type[0]+action_type[1]+action_type[2]+action_type[3]+action_type[4]+action_type[5]
    res_df['action_type_0_sigle'] = action_type[0] / action_type_sum
    res_df['action_type_1_sigle'] = action_type[1] / action_type_sum
    res_df['action_type_2_sigle'] = action_type[2] / action_type_sum
    res_df['action_type_3_sigle'] = action_type[3] / action_type_sum
    res_df['action_type_4_sigle'] = action_type[4] / action_type_sum
    res_df['action_type_5_sigle'] = action_type[5] / action_type_sum
    #     res_df['action_type_sum'] = action_type_sum
    res_df[['action_type_0','action_type_1','action_type_2','action_type_3','action_type_4','action_type_5']] = action_type[[0,1,2,3,4,5]]
    
    res_df['is_author'] = res_df['user_id'].apply(lambda x: 1 if x in authors else 0)
    
    return res_df.fillna(0)

In [3]:
# 获取最近一天的点击数量
def getCntOfOneDay(data):
    return sum(data[(data.day==0)]['cnt'].values)

# 获取最近一段时间内的点击数量
def getCntOfSomeDay(data, day_len=7):
    return sum(data[(data.day>=0) & (data.day<day_len)]['cnt'].values)

# 获取最近一段时间内的点击数量（改进版，可以自由选择col统计）
def getCntOfSomeDayWithCol(data, day_len=1, col='page'):
    return sum(data[(data.day>=0) & (data.day<day_len)][col].values)

In [4]:
# 通过video 、 act 、 register 、 lauch来构造特征
def constructDataFeature(data_url):
    train_login = pd.read_csv(data_url+'train_login.csv')
    train_act = pd.read_csv(data_url+'train_act.csv')
    train_video = pd.read_csv(data_url+'train_video.csv')
    
    # register data
    train_user = pd.read_csv(data_url+'train_users.csv')
    feature = train_user
    max_day = max(feature['register_day'])
    feature['register_day'] = max_day - feature.register_day  # day改成时间窗口截点距离
#     feature['register_day_diff'] = max_day - feature.register_day  # day改成时间窗口截点距离
    
    # login data
    train_login_feas = getCountFeature(train_login, 'login')
    feature = pd.merge(feature, train_login_feas, on='user_id', how='left')
    
    # video data
    train_video_feas = getCountFeature(train_video, 'video')
    feature = pd.merge(feature, train_video_feas, on='user_id', how='left')
    
    # act data
    train_act_feas = getCountFeatureAboutAct(train_act, 'act')
    feature = pd.merge(feature, train_act_feas, on='user_id', how='left')
    
#     feature[ [c for c in feature.columns if 'day_diff' in c ] ] = feature[ [c for c in feature.columns if 'day_diff' in c ] ].fillna(-1)
    
    # 缺失值全部补NAN
    return feature

# 通过TEST未来几天内的活跃用户来给Train集标签
def getTrainLabel(data_url, data):
    # get activity label of train from test_dataset
    train_label = []
    activity_users = getActivityUsers(data_url)
    for u in data['user_id']:
        if u in activity_users:
            train_label.append(1)
        else:
            train_label.append(0)
    data['label'] = train_label
    return data
    

In [5]:
def getDataProgram():
    start = time.time()
    print("----------------构造训练集-------------------")
    
    data_url = '../Data/data1/'
    data1 = constructDataFeature(data_url)
    data1 = getTrainLabel(data_url, data1)
    print("---------第一组数据集处理完成----------------")
    data1.to_csv(data_url+'data1.csv',index=False)
    end = time.time()
    print (end - start)
    
    data_url = '../Data/data2/'
    data2 = constructDataFeature(data_url)
    data2 = getTrainLabel(data_url, data2)
    print("---------第二组数据集处理完成----------------")
    data2.to_csv(data_url+'data2.csv',index=False)
    end = time.time()
    print (end - start)
    
    data_url = '../Data/data3/'
    data3 = constructDataFeature(data_url)
    print("---------第三组数据集处理完成----------------")
    data3.to_csv(data_url+'data3.csv',index=False)
    
    print("--------------------END----------------------")
    end = time.time()
    print (end - start)

In [6]:
getDataProgram()

----------------构造训练集-------------------
---------第一组数据集处理完成----------------
47.5337188243866
---------第二组数据集处理完成----------------
127.23727750778198
---------第三组数据集处理完成----------------
--------------------END----------------------
237.9376094341278


# 训练模型并进行预测

In [7]:
import lightgbm as lgb
from sklearn import model_selection
from sklearn.metrics import f1_score
import matplotlib.pyplot as plt

In [8]:
data1_in = pd.read_csv('../Data/data1/data1.csv')
data2_in = pd.read_csv('../Data/data2/data2.csv')
data3_in = pd.read_csv('../Data/data3/data3.csv')

# merge特征

In [9]:
# merge feature function
def mergeFeatureToData(path, data1_in, data2_in, data3_in, col='user_id'):
    feature_data1 = pd.read_csv(path + 'data1.csv')
    feature_data2 = pd.read_csv(path + 'data2.csv')
    feature_data3 = pd.read_csv(path + 'data3.csv')
    
    data1_in = pd.merge(data1_in, feature_data1, on=col, how='left')
    data2_in = pd.merge(data2_in, feature_data2, on=col, how='left')
    data3_in = pd.merge(data3_in, feature_data3, on=col, how='left')
    return data1_in, data2_in, data3_in

In [10]:
print (data1_in.shape)

path = '../Data/cnt_diff_data/'
data1_in, data2_in, data3_in = mergeFeatureToData(path, data1_in, data2_in, data3_in)
# path = '../Data/day_diff_data/'
# data1_in, data2_in, data3_in = mergeFeatureToData(path, data1_in, data2_in, data3_in)
path = '../Data/a_va_set_cnt_data/'
data1_in, data2_in, data3_in = mergeFeatureToData(path, data1_in, data2_in, data3_in)
# path = '../Data/retention_data/'
# data1_in, data2_in, data3_in = mergeFeatureToData(path, data1_in, data2_in, data3_in, col='register_day')

print (data1_in.shape)

(22382, 124)
(22382, 144)


In [6]:
# data1_in.to_csv('..\Data\out_data\data1.csv',index=False)
# data2_in.to_csv('..\Data\out_data\data2.csv',index=False)
# data3_in.to_csv('..\Data\out_data\data3.csv',index=False)

In [32]:
# Label Encoder Of device
cates = pd.Categorical(data3_in['device_type'])
categories = cates.categories
data3_in['device_type'] = cates.codes
data1_in['device_type'] = data1_in['device_type'].apply(lambda x:categories.get_loc(x))
data2_in['device_type'] = data2_in['device_type'].apply(lambda x:categories.get_loc(x))

In [5]:
def searchDaKa():
    user_act = pd.read_csv('../Data/raw_data/user_activity_log.txt', header=None,names=['user_id','day','page','video_id','author_id','action_type'],sep='\t') # 用户行为日志数据加载
    authors = user_act.groupby(['author_id']).apply(lambda x:x.shape[0])
    author_df = authors.reset_index()
    return author_df[author_df[0]>1204]['author_id'].values

daka = searchDaKa()
data1_in['daka'] = data1_in['user_id'].apply(lambda x: 1 if x in daka else 0)
data2_in['daka'] = data2_in['user_id'].apply(lambda x: 1 if x in daka else 0)
data3_in['daka'] = data3_in['user_id'].apply(lambda x: 1 if x in daka else 0)

----------------------------------------------------------------------------------------------

In [ ]:
# 列举所有的columns
for i in range(1,int(len(data2_in.columns)/5)+2):
    print (list(data2_in.columns)[(i-1)*5: i*5])

In [11]:
# 手动标定不需要的cols
drop_cols = ['login_sum','login_max','login_var','login_mean','login_3_cnt',
             'login_2_cnt','login_7_cnt',
             'device_map',
             'page_sum', 'action_type_sum']

In [12]:
# data1 = data1_in[[c for c in data1_in.columns if c not in drop_cols and c in select_cols]]
# data2 = data2_in[[c for c in data2_in.columns if c not in drop_cols and c in select_cols]]
# data3 = data3_in[[c for c in data3_in.columns if c not in drop_cols and c in select_cols]]

data1 = data1_in[[c for c in data1_in.columns if c not in drop_cols]]
data2 = data2_in[[c for c in data2_in.columns if c not in drop_cols]]
data3 = data3_in[[c for c in data3_in.columns if c not in drop_cols]]

In [13]:
print (data1.shape)
print (data2.shape)
print (data3.shape)

(22382, 138)
(26468, 138)
(51480, 137)


In [14]:
# 目前线上的参数，调过
# LGBM = lgb.LGBMClassifier(  max_depth=7,
#                             n_estimators = 220,
#                             learning_rate =0.1,     
#                             objective = 'binary',
#                             num_leaves=20,
#                             boosting_type = 'dart',
#                             feature_fraction=0.5,
#                             lambda_l1=1,
#                             lambda_l2=0.1,
#                             subsample=0.7,
#                             max_bin=25,
#                             min_data_in_leaf=20
# )
LGBM = lgb.LGBMClassifier(  max_depth=6,
                            n_estimators = 280,
                            learning_rate =0.05,     
                            objective = 'binary',
                            num_leaves=25,
                            boosting_type = 'dart',
                            feature_fraction=0.5,
                            lambda_l1=1,
                            lambda_l2=0.5,
                            subsample=0.7
)

In [15]:
# 题目规定线下 F1计算方法
def sroceF1(pred, real):
    M = set(pred)
    N = set(real)
    Precision = len(M.intersection(N))/len(M)
    Recall = len(M.intersection(N))/len(N)
    F1 = 2*Precision*Recall/(Precision+Recall)

    print("Precision=",Precision,"| Recall=",Recall)
    print("F1=",F1)

In [16]:
# 训练模型，做预测
def buildModelAndPredict(isOnLine=True, isTest=False, yuzhi=0.4, model=LGBM):
    # 线上预测
    # @return: 返回活跃用户
    if (isOnLine):
        train = pd.concat([data1, data2])
        test = data3.copy()
        train.pop('user_id')
        label = train.pop('label')
        
        model.fit(train, label)
        user_list = test.pop('user_id')
        user_df = pd.DataFrame(user_list)
        user_df['pre_act'] = model.predict_proba(test)[:,1]
        return user_df

    # 线下训练调试
    else: 
        train = data1.copy()
        test = data2.copy()
        train.pop('user_id')
        train_df_label = train.pop('label')
        train_df = train
        
        real_user = test[test.label==1]['user_id']
        user_list = test.pop('user_id')
        test.pop('label')
        test_df = test
        
        user_df = pd.DataFrame(user_list)
        model.fit(train_df, train_df_label)
        user_df['pre_act'] = model.predict_proba(test_df)[:,1]
        
        user_pre = user_df[user_df.pre_act>yuzhi]['user_id']
        sroceF1(user_pre, real_user)
        return None

In [17]:
for i in np.arange(0.39, 0.42, 0.002):
    user_df = buildModelAndPredict(isOnLine=False, isTest=False, yuzhi=i, model=LGBM)
    print(i)

Precision= 0.76398335864756 | Recall= 0.8580434621375065
F1= 0.8082861734087892
0.39
Precision= 0.7652081816376515 | Recall= 0.8573759549061781
F1= 0.8086743616649178
0.392
Precision= 0.76576277928794 | Recall= 0.8566342802047022
F1= 0.8086536441923966
0.394
Precision= 0.7665426521392505 | Recall= 0.8557442705629311
F1= 0.8086910811284388
0.396
Precision= 0.7673303589265499 | Recall= 0.8546317585107172
F1= 0.8086315789473684
0.398
Precision= 0.7680656569026489 | Recall= 0.853741748868946
F1= 0.808640674394099
0.4
Precision= 0.7689786636345395 | Recall= 0.8527034042868797
F1= 0.8086797495955546
0.402
Precision= 0.7696124446828483 | Recall= 0.8512942223540755
F1= 0.8083952530196853
0.404
Precision= 0.7700107440236369 | Recall= 0.850478380182452
F1= 0.8082466960352422
0.406
Precision= 0.7710843373493976 | Recall= 0.8496625380108285
F1= 0.808468595624559
0.408
Precision= 0.7716269975052256 | Recall= 0.8487725283690574
F1= 0.8083633538178993
0.41
Precision= 0.7728900601391986 | Recall= 0.84

In [18]:
user_df = buildModelAndPredict(isOnLine=True, isTest=False, model=LGBM)

In [19]:
LGBM.feature_importances_

array([ 79, 194, 422,  43, 300, 234, 149,  52,  65, 178, 460,  68,  54,
        39,  21,  34,  12,  39,  16,  13,  20,   6,  20,  13,   0,  56,
       193, 130,  43,  96,  86,  53,  60,  76,  27,  21,  53,  32,  11,
        71,  23,  10,   2,  34,  16,  23,  12,  15,   8,  23,   6,   5,
         1,  46,  13,  37,  17,  26,  11,  33,  11,  11,   4,  59,  63,
        47,  92,  39,  63,  95,  29,  45,   7,  85,   8,   2,   0,   0,
         0,  50,  21,  12,   5,   6,   5,   0,   0,   0,   0,   0,   0,
        70,  36,  17,   6,   8,   8,   4,   0,   0,   0,   0,   0,  77,
        97,  64,  37,   0,   0, 134,  23,   9,  12,   0,   0,   2, 171,
       114,  94,  77,  29,  11, 138,  70,  86,  95,  49,  37,  63,  44,
        52,  27,  39,  48, 114,  99])

In [20]:
# user_pre = user_df.sort_values(by='pre_act', axis = 0, ascending = False)['user_id'].head(24500)
user_pre = user_df[user_df.pre_act>0.4]['user_id']

In [21]:
len(user_pre)

24379

In [18]:
user_df.to_csv(r'..\Data\out_data\user_out.csv',index=False)

# 结果数据提交

In [22]:
user_pre.to_csv('../Output/0627_136_808640_24379_4.csv', index=False)

// 复现82Line
116_80236_24870_40 线上：0.820298

// 82Line +resiter_day原数据 fillna（0）
117_80263_24890_40 线上：0.819631

// 82Line -page_sum和action_sum fillna（0）
114_8019_24781_40  线上：0.820071

// 82Line +day_diff
134_80236_24991_40 线上：0.820108

// 82Line -page_sum和action_sum,不作fillna
114_8024099_24712_404  线上：0.820083         0.4:0.80188; 0.404:0.8024099

0624:

// +cnt_diff
0624_126_80767_24746_4 线上：0.820209                  0.4:0.80767

// +day_diff
0624_132_802742_24959_4  线上：0.819880              0.4:0.802742; 0.404:0.802946

// +retention
0624_124_802092_25073_4  线上: 0.819508              0.4:0.802092; 0.416:0.8027104

// +a_va_set_cnt_data
0624_122_80156_24954_4  线上:  0.820600            0.4:0.801560; 0.412:0.8024499

// +all
0624_162_8066504_24906_4 线上: 0.820019            0.4:0.80665040; 0.414:0.778059

// +all 调参
0626_162_807501_24624_4 线上:             0.4:0.807501; 0.394:0.807855

# 模型融合

In [2]:
out_data1 = pd.read_csv(r'..\Data\out_data\xgb_result_0625.csv')
out_data2 = pd.read_csv(r'..\Data\out_data\user_out.csv')

In [3]:
out_user = pd.merge(out_data1, out_data2, on='user_id',how='left')

In [4]:
out_user['score'] = out_user['label']*0.4 + out_user['pre_act']*0.6

In [9]:
users = out_user[out_user.score>0.4]

In [10]:
len(users)

24646

In [11]:
users.to_csv(r'../Output/rohe_24646_4.csv', index=False)